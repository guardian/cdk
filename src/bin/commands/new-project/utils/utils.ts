import camelCase from "lodash.camelcase";
import upperFirst from "lodash.upperfirst";

export const pascalCase = (str: string): string => {
  return upperFirst(camelCase(str));
};

export interface Name {
  kebab: string;
  pascal: string;
}
