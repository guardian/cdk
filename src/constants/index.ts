export * from "./access";
export * from "./context-keys";
export * from "./library-info";
export * from "./regex-pattern";
export * from "./ssm-parameter-paths";
export * from "./tag-keys";
export * from "./tracking-tag";
