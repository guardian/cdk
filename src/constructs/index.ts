export * as acm from "./acm";
export * as autoscaling from "./autoscaling";
export * as cloudwatch from "./cloudwatch";
export * as core from "./core";
export * as dns from "./dns";
export * as dynamodb from "./dynamodb";
export * as ec2 from "./ec2";
export * as ecs from "./ecs";
export * as iam from "./iam";
export * as kinesis from "./kinesis";
export * as lambda from "./lambda";
export * as loadbalancing from "./loadbalancing";
export * as rds from "./rds";
export * as s3 from "./s3";
export * as ses from "./ses";
export * as vpc from "./vpc";
export * as root from "./root";
