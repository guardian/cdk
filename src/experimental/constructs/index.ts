export * from "./error-budget-alarm";
export * from "./root";
