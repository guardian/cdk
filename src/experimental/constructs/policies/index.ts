export * from "./kinesis-put-records";
