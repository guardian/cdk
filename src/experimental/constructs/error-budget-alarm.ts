import { Duration } from "aws-cdk-lib";
import type { CfnCompositeAlarm, IMetric } from "aws-cdk-lib/aws-cloudwatch";
import {
  Alarm,
  AlarmRule,
  AlarmState,
  ComparisonOperator,
  CompositeAlarm,
  MathExpression,
  TreatMissingData,
} from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import type { ITopic } from "aws-cdk-lib/aws-sns";
import { Topic } from "aws-cdk-lib/aws-sns";
import { Construct } from "constructs";
import type { GuStack } from "../../constructs/core";

export interface ErrorBudgetAlarmProps {
  sloName: string;
  sloTarget: number;
  badEvents: IMetric;
  validEvents: IMetric;
  snsTopicNameForAlerts: string;
}

interface BurnRate {
  speed: "Fast" | "Medium" | "Slow";
  burnRate: number;
}

interface BurnRateMonitoring extends BurnRate {
  longPeriod: Duration;
  shortPeriod: Duration;
}

interface MonitorBurnRateForPeriodProps {
  alarmName: string;
  burnRateSpeed: BurnRate;
  period: Duration;
  errorBudget: number;
  badEvents: IMetric;
  validEvents: IMetric;
}

interface CompositeBurnRateAlarmProps {
  sloName: string;
  burnRateMonitoring: BurnRateMonitoring;
  errorBudget: number;
  badEvents: IMetric;
  validEvents: IMetric;
  snsTopic: ITopic;
  suppressorAlarm?: CompositeBurnRateAlarm;
}

/**
 * Use this construct to create alarms that will notify you when you are at risk of missing your SLO targets.
 *
 * These alarms follow the recommendations in the [SRE Workbook](https://sre.google/workbook/alerting-on-slos/) to
 * implement Multiwindow, Multi-Burn-Rate Alerts.
 *
 * The CloudWatch implementation of this strategy creates nine separate CloudWatch alarms. The main alarms that users
 * need to care about are the three composite alarms. These track different burn rates of error budget consumption:
 * fast (e.g. complete outages), medium (e.g. moderate error percentage over hours, rather than minutes) and slow (e.g.
 * a small percentage of errors sustained over a much longer period). Multiple burn rates are used to ensure that
 * we are able to pick up all scenarios which could threaten the SLO, without interrupting developers with an
 * unnecessary level of urgency.
 *
 * Each composite alarm creates two child alarms; these are used to support multiple windows (or periods in CloudWatch
 * terminology). A composite alarm is configured to send notifications if both of its child alarms are in an alarm state.
 * The child alarms themselves have no actions/notifications configured and users should not need to interact with them
 * directly. Multiple windows are helpful for accuracy and reset time. The long window helps us to avoid sending an
 * alert in scenarios where no intervention is required, for example a very brief spike in errors that is resolved
 * automatically (e.g. when AWS replaces a 'bad' EC2 instance). The short window helps to ensure that the alert is reset
 * (or moved back into OK state, in CloudWatch terminology) promptly once a problem has been resolved.
 *
 * The composite alarms have an awareness of priority, meaning that low priority alarm notifications are suppressed if
 * a higher priority alarm is already firing. That is, if you receive an alert about fast error budget consumption, you
 * should not receive an alert from the medium or slow versions of the alarm.
 */
export class GuErrorBudgetAlarmExperimental extends Construct {
  constructor(scope: GuStack, props: ErrorBudgetAlarmProps) {
    super(scope, props.sloName); // The assumption here is that `sloName` will be unique (per stack)

    const errorBudget = 1 - props.sloTarget;

    const snsTopic: ITopic = Topic.fromTopicArn(
      scope,
      `SnsSloAlarmsFor${props.sloName}`,
      `arn:aws:sns:${scope.region}:${scope.account}:${props.snsTopicNameForAlerts}`
    );

    const fastBurnRate: BurnRateMonitoring = {
      speed: "Fast",
      burnRate: 14.4,
      longPeriod: Duration.hours(1),
      shortPeriod: Duration.minutes(5),
    };

    const fastBurnAlarm = new CompositeBurnRateAlarm(scope, {
      sloName: props.sloName,
      badEvents: props.badEvents,
      burnRateMonitoring: fastBurnRate,
      errorBudget,
      snsTopic,
      validEvents: props.validEvents,
    });

    const mediumBurnRateMonitoring: BurnRateMonitoring = {
      speed: "Medium",
      burnRate: 6,
      longPeriod: Duration.hours(6),
      shortPeriod: Duration.minutes(30),
    };

    const mediumBurnRateAlarm = new CompositeBurnRateAlarm(scope, {
      sloName: props.sloName,
      badEvents: props.badEvents,
      burnRateMonitoring: mediumBurnRateMonitoring,
      errorBudget,
      snsTopic,
      validEvents: props.validEvents,
      suppressorAlarm: fastBurnAlarm,
    });

    const slowBurnRate: BurnRateMonitoring = {
      speed: "Slow",
      burnRate: 3,
      longPeriod: Duration.days(1),
      shortPeriod: Duration.hours(2),
    };

    new CompositeBurnRateAlarm(scope, {
      sloName: props.sloName,
      badEvents: props.badEvents,
      burnRateMonitoring: slowBurnRate,
      errorBudget,
      // TODO - use a different topic for tickets or figure out how to route notifications vs tickets using a single topic
      snsTopic,
      validEvents: props.validEvents,
      suppressorAlarm: mediumBurnRateAlarm,
    });
  }
}

class MonitorBurnRateForPeriod extends Alarm {
  constructor(scope: GuStack, props: MonitorBurnRateForPeriodProps) {
    const undesirableErrorBudgetConsumption = props.errorBudget * props.burnRateSpeed.burnRate;
    super(scope, props.alarmName, {
      alarmName: props.alarmName,
      metric: new MathExpression({
        expression: "badEvents/validEvents",
        label: "Observed failure rate",
        usingMetrics: {
          badEvents: props.badEvents,
          validEvents: props.validEvents,
        },
        period: props.period,
      }),
      evaluationPeriods: 1,
      threshold: undesirableErrorBudgetConsumption,
      comparisonOperator: ComparisonOperator.GREATER_THAN_THRESHOLD,
      treatMissingData: TreatMissingData.NOT_BREACHING,
    });
  }
}

class CompositeBurnRateAlarm extends CompositeAlarm {
  constructor(scope: GuStack, props: CompositeBurnRateAlarmProps) {
    const alarmName = `${props.sloName}${props.burnRateMonitoring.speed}BurnCompositeAlarm`;
    const sharedIndividualAlarmProps = {
      badEvents: props.badEvents,
      burnRateSpeed: {
        speed: props.burnRateMonitoring.speed,
        burnRate: props.burnRateMonitoring.burnRate,
      },
      errorBudget: props.errorBudget,
      validEvents: props.validEvents,
    };
    super(scope, alarmName, {
      alarmRule: AlarmRule.allOf(
        AlarmRule.fromAlarm(
          new MonitorBurnRateForPeriod(scope, {
            ...sharedIndividualAlarmProps,
            alarmName: `ChildAlarmLongPeriod${alarmName}`,
            period: props.burnRateMonitoring.longPeriod,
          }),
          AlarmState.ALARM
        ),
        AlarmRule.fromAlarm(
          new MonitorBurnRateForPeriod(scope, {
            ...sharedIndividualAlarmProps,
            alarmName: `ChildAlarmShortPeriod${alarmName}`,
            period: props.burnRateMonitoring.shortPeriod,
          }),
          AlarmState.ALARM
        )
      ),
    });
    this.addAlarmAction(new SnsAction(props.snsTopic));
    if (props.suppressorAlarm) {
      const cfnAlarm = this.node.defaultChild as CfnCompositeAlarm;
      cfnAlarm.actionsSuppressor = props.suppressorAlarm.alarmArn;
      cfnAlarm.actionsSuppressorWaitPeriod = 120;
      cfnAlarm.actionsSuppressorExtensionPeriod = props.burnRateMonitoring.shortPeriod.toSeconds();
    }
  }
}
