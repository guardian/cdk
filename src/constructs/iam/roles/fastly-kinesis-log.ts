import { AccountPrincipal } from "aws-cdk-lib/aws-iam";
import { FASTLY_AWS_ACCOUNT_ID } from "../../../constants/fastly-aws-account-id";
import type { GuStack } from "../../core";
import { GuFastlyCustomerIdParameter } from "../../core";
import type { GuKinesisStream } from "../../kinesis";
import { GuKinesisPutRecordsPolicy } from "../policies/kinesis-put-records";
import { GuRole } from "./roles";

export interface GuFastlyKinesisLogRoleProps {
  /**
   * The Kinesis stream into which Fastly will put records
   */
  stream: GuKinesisStream;
  /**
   * Name of the IAM role
   */
  roleName?: string;
}

/**
 * A construct to create an IAM Role for Fastly to assume in order to write to a
 * specific Kinesis stream.
 *
 * In order to use this construct, an SSM parameter named `/account/external/fastly/customer.id`
 * needs to exist in the AWS account's parameter store, and the value should be
 * the Guardian's Fastly customer id.
 *
 */
export class GuFastlyKinesisLogRole extends GuRole {
  constructor(scope: GuStack, id: string, props: GuFastlyKinesisLogRoleProps) {
    const fastlyCustomerId = GuFastlyCustomerIdParameter.getInstance(scope).valueAsString;

    super(scope, id, {
      roleName: props.roleName,
      assumedBy: new AccountPrincipal(FASTLY_AWS_ACCOUNT_ID),
      externalIds: [fastlyCustomerId],
    });

    const policy = new GuKinesisPutRecordsPolicy(scope, "GuKinesisPutRecordsPolicy", {
      stream: props.stream,
    });

    policy.attachToRole(this);
  }
}
