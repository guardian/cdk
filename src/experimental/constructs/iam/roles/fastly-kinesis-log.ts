import { AccountPrincipal } from "aws-cdk-lib/aws-iam";
import { FASTLY_AWS_ACCOUNT_ID } from "../../../../constants/fastly-aws-account-id";
import type { GuStack } from "../../../../constructs/core";
import { GuFastlyCustomerIdParameter } from "../../../../constructs/core";
import { GuRole } from "../../../../constructs/iam";
import type { GuKinesisStream } from "../../../../constructs/kinesis";
import { GuKinesisPutRecordsPolicyExperimental } from "../../policies/kinesis-put-records";

export interface GuFastlyKinesisLogRoleProps {
  /**
   * The Kinesis stream into which Fastly will put records
   */
  stream: GuKinesisStream;
  /**
   * The name of the IAM role
   */
  roleName?: string;
  /**
   * The name of the policy attached to this role which allows writing to the Kinesis stream
   */
  policyName?: string;
}

/**
 * A construct to create an IAM Role for Fastly to assume in order to write to a
 * specific Kinesis stream.
 *
 * In order to use this construct, an SSM parameter named `/account/external/fastly/customer.id`
 * needs to exist in the AWS account's parameter store, and the value should be
 * the Guardian's Fastly customer id.
 *
 */
export class GuFastlyKinesisLogRoleExperimental extends GuRole {
  constructor(scope: GuStack, id: string, props: GuFastlyKinesisLogRoleProps) {
    const fastlyCustomerId = GuFastlyCustomerIdParameter.getInstance(scope).valueAsString;
    const { policyName, roleName, stream } = props;

    super(scope, id, {
      roleName,
      assumedBy: new AccountPrincipal(FASTLY_AWS_ACCOUNT_ID),
      externalIds: [fastlyCustomerId],
    });

    const policy = new GuKinesisPutRecordsPolicyExperimental(
      scope,
      policyName ?? "GuKinesisPutRecordsPolicyExperimental",
      {
        stream,
      },
    );

    policy.attachToRole(this);
  }
}
