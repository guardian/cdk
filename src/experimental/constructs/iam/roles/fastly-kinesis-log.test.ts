import { Template } from "aws-cdk-lib/assertions";
import { FASTLY_AWS_ACCOUNT_ID } from "../../../../constants";
import { GuKinesisStream } from "../../../../constructs/kinesis";
import { simpleGuStackForTesting } from "../../../../utils/test";
import { GuFastlyKinesisLogRoleExperimental } from "./fastly-kinesis-log";

describe("The GuFastlyKinesisLogRole construct", () => {
  it("correctly wires up the policy", () => {
    const stack = simpleGuStackForTesting();
    const testStream = new GuKinesisStream(stack, "testStream");
    new GuFastlyKinesisLogRoleExperimental(stack, "testKinesisLogRole", {
      stream: testStream,
      roleName: "writeToKinesisRoleTest",
    });
    expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
  });

  it("assumes the correct role", () => {
    const stack = simpleGuStackForTesting();
    const testStream = new GuKinesisStream(stack, "testStream");
    new GuFastlyKinesisLogRoleExperimental(stack, "testKinesisLogRole", {
      stream: testStream,
      roleName: "writeToKinesisRoleTest",
    });
    Template.fromStack(stack).hasResourceProperties("AWS::IAM::Role", {
      AssumeRolePolicyDocument: {
        Version: "2012-10-17",
        Statement: [
          {
            Action: "sts:AssumeRole",
            Condition: {
              StringEquals: {
                "sts:ExternalId": {
                  Ref: "FastlyCustomerId",
                },
              },
            },
            Effect: "Allow",
            Principal: {
              AWS: {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      Ref: "AWS::Partition",
                    },
                    `:iam::${FASTLY_AWS_ACCOUNT_ID}:root`,
                  ],
                ],
              },
            },
          },
        ],
      },
    });
  });

  it("allows more than one role in the stack", () => {
    const stack = simpleGuStackForTesting();
    const testStreamOne = new GuKinesisStream(stack, "testStreamOne");
    const testStreamTwo = new GuKinesisStream(stack, "testStreamTwo");

    new GuFastlyKinesisLogRoleExperimental(stack, "testKinesisLogRoleOne", {
      stream: testStreamOne,
      policyName: "writeToTestStreamPolicyOne",
      roleName: "writeToKinesisRoleTestOne",
    });

    new GuFastlyKinesisLogRoleExperimental(stack, "testKinesisLogRoleTwo", {
      stream: testStreamTwo,
      policyName: "writeToTestStreamPolicyTwo",
      roleName: "writeToKinesisRoleTestTwo",
    });

    // Assert that two separate roles exist in the stack with policies that allow writing to Kinesis
    Template.fromStack(stack).resourceCountIs("AWS::IAM::Role", 2);

    Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
      PolicyDocument: {
        Statement: [
          {
            Action: ["kinesis:PutRecords", "kinesis:ListShards"],
            Effect: "Allow",
            Resource: {
              "Fn::GetAtt": ["testStreamOneE597ADBB", "Arn"],
            },
          },
        ],
        Version: "2012-10-17",
      },
      PolicyName: "writeToTestStreamPolicyOneDD9816F1",
      Roles: [
        {
          Ref: "testKinesisLogRoleOne47025733",
        },
      ],
    });

    Template.fromStack(stack).hasResourceProperties("AWS::IAM::Policy", {
      PolicyDocument: {
        Statement: [
          {
            Action: ["kinesis:PutRecords", "kinesis:ListShards"],
            Effect: "Allow",
            Resource: {
              "Fn::GetAtt": ["testStreamTwoBFC0E912", "Arn"],
            },
          },
        ],
        Version: "2012-10-17",
      },
      PolicyName: "writeToTestStreamPolicyTwo61962E93",
      Roles: [
        {
          Ref: "testKinesisLogRoleTwoC08DB6A6",
        },
      ],
    });
  });
});
