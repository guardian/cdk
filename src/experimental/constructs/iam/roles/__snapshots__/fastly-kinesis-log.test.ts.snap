// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuFastlyKinesisLogRole construct correctly wires up the policy 1`] = `
{
  "Conditions": {
    "AwsCdkKinesisEncryptedStreamsUnsupportedRegions": {
      "Fn::Or": [
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-north-1",
          ],
        },
        {
          "Fn::Equals": [
            {
              "Ref": "AWS::Region",
            },
            "cn-northwest-1",
          ],
        },
      ],
    },
  },
  "Metadata": {
    "gu:cdk:constructs": [
      "GuStack",
      "GuKinesisStream",
      "GuFastlyCustomerIdParameter",
      "GuFastlyKinesisLogRoleExperimental",
      "GuKinesisPutRecordsPolicyExperimental",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "FastlyCustomerId": {
      "Default": "/account/external/fastly/customer.id",
      "Description": "SSM parameter containing the Fastly Customer ID. Can be obtained from https://manage.fastly.com/account/company by an admin",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "GuKinesisPutRecordsPolicy28D5CD2D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:PutRecords",
                "kinesis:ListShards",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "testStream8BCA7523",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GuKinesisPutRecordsPolicy28D5CD2D",
        "Roles": [
          {
            "Ref": "testKinesisLogRole5E3B33EE",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testKinesisLogRole5E3B33EE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Condition": {
                "StringEquals": {
                  "sts:ExternalId": {
                    "Ref": "FastlyCustomerId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::717331877981:root",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "writeToKinesisRoleTest",
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testStream8BCA7523": {
      "Properties": {
        "RetentionPeriodHours": 24,
        "ShardCount": 1,
        "StreamEncryption": {
          "Fn::If": [
            "AwsCdkKinesisEncryptedStreamsUnsupportedRegions",
            {
              "Ref": "AWS::NoValue",
            },
            {
              "EncryptionType": "KMS",
              "KeyId": "alias/aws/kinesis",
            },
          ],
        },
        "Tags": [
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/cdk",
          },
          {
            "Key": "Stack",
            "Value": "test-stack",
          },
          {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::Kinesis::Stream",
    },
  },
}
`;
